cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageService.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.metadata.lineage;
+
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Collections2;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Service to compute Lineage based on Dataset accesses of a Program stored in {@link LineageDataset}.
+ */
+public class LineageService {
+  private static final Logger LOG = LoggerFactory.getLogger(LineageService.class);
+
+  private final LineageDataset lineageDataset;
+
+  public LineageService(LineageDataset lineageDataset) {
+    this.lineageDataset = lineageDataset;
+  }
+
+  public Lineage computeLineage(final Id.DatasetInstance sourceDataset, long start, long end, int levels) {
+    LOG.trace("Computing lineage for dataset {}, start {}, end {}, levels {}", sourceDataset, start, end, levels);
+    Set<Relation> relations = new HashSet<>();
+    Set<Id.DatasetInstance> visitedDatasets = new HashSet<>();
+    Set<Id.DatasetInstance> toVisitDatasets = new HashSet<>();
+    Set<Id.Program> visitedPrograms = new HashSet<>();
+    Set<Id.Program> toVisitPrograms = new HashSet<>();
+
+    toVisitDatasets.add(sourceDataset);
+    for (int i = 0; i < levels; ++i) {
```

Comment:
This is a pretty complicated logic, better break it down by intention. E.g (simplified).

```java
// Compute lineage progressively for iteration
for (int i = 0; i < levels; i++) {
  toVisitPrograms.clear();
  gatherPrograms(visitedDatasets, toVisitDatasets, toVisitPrograms);
  toVisitDatasets.clear();
  gatherLineageDatasets(visitedDatasets, toVisitPrograms, toVisitDatasets);
}